flycheck.el

```diff
@@ -6976,6 +6976,45 @@ Otherwise return the previously used cache directory."
         (or flycheck-haskell-ghc-cache-directory
             (make-temp-file "flycheck-haskell-ghc-cache" 'directory))))
 
+(defun flycheck-haskell--find-default-directory (checker)
+  "Come up with a sutable default directory for Haskell to run
+checker in.
+
+In case of `haskell-stack-ghc' checker it is directory with
+stack.yaml file. If there's no stack.yaml file in any parent
+directory, it will be the directory that \"stack path --project-root\"
+command returs.
+
+For all other checkers, it is the closest parent directory that
+contains cabal file."
+  (pcase checker
+    (`haskell-stack-ghc
+     (or
+      (locate-dominating-file (buffer-file-name) "stack.yaml")
+      (when (executable-find "stack")
+        (let* ((stack-output
+                (process-lines "stack" "path" "--project-root"))
+               (stack-dir
+                (when stack-output
+                  (car stack-output))))
+          (when (and stack-dir
+                     (file-directory-p stack-dir))
+            stack-dir)))))
+    (_
+     (if (fboundp 'haskell-cabal-find-file)
+         (let ((cabal-file (haskell-cabal-find-file)))
+           (when cabal-file
+             (file-name-directory cabal-file)))
+       (locate-dominating-file (file-name-directory (buffer-file-name))
+                               (lambda (dir)
+                                 (not
+                                  (null
+                                   (directory-files dir
+                                                    nil ;; full
+                                                    ".+\\.cabal\\'"
+                                                    t ;; nosort
+                                                    )))))))))
+
```

Comment:
I think `haskell-cabal-find-file` doesn't add much beyond stopping upwards traversal if file/directory ownership changes. Which sounds questionable to me anyway. Probably best to avoid referring to anything in haskell-mode here IMO.